### PR TITLE
Plotting Wrappers: Occupancy Histogram

### LINKED ISSUE #5: Plot occupancy heatmap
See example implementation [here](https://github.com/pyratlib/pyrat/blob/main/pyratlib/processing.py)

### LINKED ISSUE #388: Plotting wrappers: Occupancy Heatmap
Sub-issue of #282. See the description provided [here](https://github.com/neuroinformatics-unit/movement/issues/282#issuecomment-2612906175).

#5 has an example and some further information + examples.

Keeping this issue separate since #5 also discusses retaining the data output, as well as the figure, which is beyond the scope of this issue.

### PR DESCRIPTION
**Pull request by willGraham01**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

See #388 and related, #5 (which is actually also closed)

**What does this PR do?**

Adds the `plot_occupancy` function to the `movement.plots` module. This function takes in `(time, space [x, y])`-data and produces a histogram showing the distribution of positions across all time-points.

By default, any additional axes in the input `da` (`DataArray`) are collapsed onto the 0th-index, to provide the expected 2D data input. The `selection` argument can be used by the user to specify alternative coordinates along non-spacetime dimensions to collapse onto instead.

`plot_occupancy` returns the usual figure and axes objects, however also returns information from the plotted histogram as its third value. This is mainly because this information is difficult to re-extract from the returned `axes` figure. The `counts` information in particular would technically otherwise be lost since `QuadMesh` objects (that store histograms) only retain the colour-mapped values (which may blur across bins with similar, but distinct counts), and not the raw counts in each bin.

## References

Closes #388. ~~Additionally, this hopefully goes some way towards addressing #5, since we are returning the histogram data as the 3rd return value.~~
Closes #5 too.

## How has this PR been tested?

Addition of tests to cover expected functionality, and possible edge cases.

## Is this a breaking change?

No

## Does this PR require an update to the documentation?

#410 

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [ ] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by willGraham01** [comment]
@sfmig your comment on #5 indicates that it would be useful to have certain bits of information from the plot as outputs from this kind of function. Currently I'm just exposing the other `hist2d` outputs (that are suppressed by the wrapper otherwise) to the user here, not sure if you had more detailed outputs in mind when writing [your comment](https://github.com/neuroinformatics-unit/movement/issues/5#issuecomment-2541937319).

But if so, we can also close #5 with this PR too.

**Comment by sfmig** [comment]
thanks for checking @willGraham01 !

The point of that comment was that often you not only want the figure, but also the 2D array with the bin counts. From your comment ...
> Currently I'm just exposing the other hist2d outputs

seems like that is covered? So I think we can close #5 yay 😄  🚀

**Comment by willGraham01** [comment]
> (Just fyi I vaguely remember this was something Sepi requested but not sure)

I hope it is b/c otherwise I've just wasted 10 hours of Niko's grant :hand_over_mouth: :joy: But will mark #5 as closable by this :partying_face:

**Comment by niksirbi** [comment]
I will finish reviewing this tomorrow,  but I can already do some cool things with this!

image

<details><summary>See source code for this figure</summary>

```python
import numpy as np
from matplotlib import pyplot as plt

from movement import sample_data
from movement.plots import plot_occupancy

# Load the sample dataset 
ds = sample_data.fetch_dataset("DLC_two-mice.predictions.csv")

# Compute the centroid of all keypoints
centroid_position = ds.position.mean("keypoints")

image = plt.imread(ds.attrs["frame_path"])

# Construct bins of size 20x20 pixels that cover the entire image
bin_pix = 30
bins = [
    np.arange(0, image.shape[0] + bin_pix, bin_pix),
    np.arange(0, image.shape[1] + bin_pix, bin_pix),
]

# Initialize the figure and axis
fig, ax = plt.subplots()

# Show the image
ax.imshow(image)

# Plot the occupancy 2D histogram for each individual
_, _, hist_data = plot_occupancy(
    da=centroid_position,
    selection={"individuals": "individual1"},
    ax=ax,
    cmap="viridis",
    alpha=0.5,
    bins=bins,
    cmin=3,      # Set the minimum shown count
    norm="log"
)

# Set the axis limits to match the image
ax.set_xlim(0, image.shape[1])
ax.set_ylim(image.shape[0], 0)
```

</details>

File: movement/plots/occupancy.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    desired, the ``return_hist_info`` argument should be provided as ``True``.
+    This will force the function to return a second output value, which is a
+    dictionary containing the bin edges and bin counts that were used to create
+    the histogram.
+
+    For data with ``N`` time-points, the dictionary output has key-value pairs;
+    - ``xedges``, an ``(N+1,)`` ``numpy`` array specifying the bin edges in the
+    first spatial dimension.
+    - ``yedges``, same as ``xedges`` but for the second spatial dimension.
+    - ``counts``, an ``(N, N)`` ``numpy`` array with the count for each bin.
```
Comment (niksirbi) [inline movement/plots/occupancy.py:71]: Something here doesn't add up. Why would he shapes fo `xedges`, `yedges` and `counts` be related to the number of time-points `N`?
We are binning over space, not time.

File: movement/plots/occupancy.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    ``counts[x, y]`` is the number of datapoints in the
+    ``(xedges[x], xedges[x+1]), (yedges[y], yedges[y+1])`` bin. These values
+    are those returned from ``matplotlib.pyplot.Axes.hist2d``.
+
+    Note that the ``counts`` values do not necessarily match the mappable
+    values that one gets from extracting the data from the
+    ``matplotlib.collections.QuadMesh`` object (that represents the rendered
+    histogram) via its ``get_array()`` attribute.
```
Comment (niksirbi) [inline movement/plots/occupancy.py:80]: I don't know if this is important enough to be mentioned here. To most people it would never occur to do that anyway (I might be wrong).

**Review by niksirbi — CHANGES_REQUESTED** [review]
Thanks @willGraham01!

I’ve added some comments, mostly about aligning the function signature (and default behavior) with that of `plot_trajectory()`.

Regarding your discussion with Sofía:  
Yes, this approach technically meets the requirement of also obtaining the occupancy data as a 2D array, which is excellent. However, it can be slightly awkward to always rely on the plotting function when all you need is the occupancy array. There may be scenarios where the user only wants the 2D occupancy array—without the plot—for comparisons with neural data. From that perspective, it might be more intuitive to have a dedicated `compute_occupancy` function that returns both the 2D array and the bin edges. We could discuss the best data structure to return—whether that’s an `xr.DataArray` or multiple NumPy arrays, similar to `hist2d`.

In any case, I suggest merging this PR with just `plot_occupancy` (after addressing my comments) and leaving `compute_occupancy` for a future PR. We just need to ensure that both functions produce consistent histogram data, i.e. `compute_occupancy` should use the same underlying method as `hist2d`.

File: movement/plots/occupancy.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    Parameters
+    ----------
+    da : xarray.DataArray
+        Spatial data to create histogram for. NaN values are dropped.
+    selection : dict[str, Hashable], optional
+        Mapping of dimension identifiers to the coordinate along that dimension
+        to plot. "time" and "space" dimensions are ignored. For example,
+        ``selection = {"individuals": "Bravo"}`` will create the occupancy
+        histogram for the individual "Bravo", instead of the occupancy
+        histogram for the 0-indexed entry on the ``"individuals"`` dimension.
```
[the remark this answers is not in the document]
Comment (stellaprins) [reply movement/plots/occupancy.py:41]: I would go for option **B** and allow multiple individuals for occupancy with as default the summing counts for all individuals. If there is multi-individual data it seems likely to me that it is desirable to quickly have an overview that includes occupancy of all animals (or groups of them). If it turns out the default is generally uninformative, the default can always be changed to the first individual while still giving the option to allow multiple individuals to be specified.

File: movement/plots/occupancy.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    # right now because we still potentially have a (time, space, individuals)
+    # array and so dropping NaNs along any axis may remove valid points for
+    # other times / individuals.
+    # Since we only care about a count, we can just unravel the individuals
+    # dimension and create a "long" array of points. For example, a (10, 2, 5)
+    # time-space-individuals DataArray becomes (50, 2).
+    if "individuals" in data.dims:
+        data = data.stack(
+            {"new": ("time", "individuals")}, create_index=False
+        ).rename({"new": "time"})
```
Comment (niksirbi) [inline movement/plots/occupancy.py:157]: In some cases I get the following warning here:

```python
/Users/nsirmpilatze/Code/NIU/movement/movement/plots/occupancy.py:157: UserWarning: rename 'new' to 'time' does not create an index anymore. Try using swap_dims instead or use set_index after rename to create an indexed coordinate.
  ).rename({"new": "time"})
```

File: movement/plots/occupancy.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+    Examples
+    --------
+    Simple use-case is to plot a histogram of the centroid of all
+    keypoints, aggregated over all individuals.
+
+    >>> from movement import sample_data
+    >>> from movement.plots import plot_occupancy
+    >>> positions = sample_data.fetch_dataset(
+    ...     "SLEAP_three-mice_Aeon_proofread.analysis.h5"
```
Comment (niksirbi) [inline movement/plots/occupancy.py:89]: I think this particualr dataset is the most confusing one for this example, because the animals contain a single keypont named "centroid". While reading the example, I had a hard time differentiating between "the centroid of given ketypints" and "the keypoints named centroid".

To save ourselves this trouble, I suggest using the "DLC_two-mice.predictions.csv" dataset instead. That once contains two individuals, with multiple keypoints each.

```suggestion
    ...     "DLC_two-mice.predictions.csv"
```

File: movement/plots/occupancy.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    >>> plot_occupancy(
+    ...     positions,
+    ...     # use only one keypoint in computation of centroid
+    ...     keypoints="centroid",
+    ...     # only aggregate for two individuals
+    ...     individuals=[
+    ...         "AEON3B_TP1",
+    ...         "AEON3B_TP2",
+    ...     ],
+    ... )
```
Comment (niksirbi) [inline movement/plots/occupancy.py:106]: If you switch to the dataset I suggested, this can be changed to:
```suggestion
    >>> plot_occupancy(
    ...     positions,
    ...     # plot the centroid of keypoints located on the head
    ...     keypoints=["snout", "leftear", "rightear"],
    ...     # only plot data for 1 individual
    ...     individuals="individual1"
    ... )
```

File: movement/plots/occupancy.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+    >>> plot_occupancy(
+    ...     positions,
+    ...     # use only one keypoint in computation of centroid
+    ...     keypoints="centroid",
+    ...     # only aggregate for two individuals
+    ...     individuals=[
+    ...         "AEON3B_TP1",
+    ...         "AEON3B_TP2",
+    ...     ],
```
Comment (niksirbi) [inline movement/plots/occupancy.py:119]: If you switch to using the dataset I proposed:
```suggestion
    >>> plot_occupancy(  
    ...     positions,  
    ...     # plot the centroid of keypoints located on the head  
    ...     keypoints=["snout", "leftear", "rightear"],  
    ...     # only plot data for 1 individual  
    ...     individuals="individual1",  
```

File: movement/plots/occupancy.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        fig, ax = plt.subplots()
+    counts, xedges, yedges, hist_image = ax.hist2d(
+        data.sel(space=x_coord), data.sel(space=y_coord), **kwargs
+    )
+    colourbar = fig.colorbar(hist_image, ax=ax)
+    colourbar.solids.set(alpha=1.0)
+
+    space_unit = data.attrs.get("space_unit", "pixels")
+    ax.set_xlabel(f"{x_coord} ({space_unit})")
+    ax.set_ylabel(f"{y_coord} ({space_unit})")
```
Comment (niksirbi) [inline movement/plots/occupancy.py:183]: For reasons mentioned in issue #417, I suggest ditching the space unit and just labelling the axes as "x" and "y".

**Review by niksirbi — APPROVED** [review]
Thanks @willGraham01! I like your approach to aggregating, it's entirely sensible.

I left some final comments to be addressed. I'll pre-approve this PR so you have the freedom to merge it as soon as you're done addressing those comments.
